### PR TITLE
[REVIEW] Add support for LimitAlarmType and CertificateExpirationAlarmType

### PR DESCRIPTION
**Pull request by opcua-tsn-team-kalycito**

### CODE REVIEW

File: src/server/ua_subscription_alarms_conditions.c [hunk 1]
```diff
@@ -2612,6 +2661,144 @@ UA_StatusCode UA_Server_deleteCondition(UA_Server *server, const UA_NodeId condi
...
+    UA_StatusCode retval = UA_Server_readObjectProperty(server, conditionId,
+                                UA_QUALIFIEDNAME(0, CONDITION_FIELD_LOWLIMIT),
+                                value);
+    *lowLimit = *(UA_Double*) value->data;
+    UA_Variant_delete(value);
+    return retval;
+}
+
+static UA_StatusCode getLowLowLimit(UA_Server *server, UA_NodeId conditionId, UA_Double *lowLowLimit) {
+    UA_Variant *value = UA_Variant_new();
```
Comment (basyskom-meerkoetter) [inline src/server/ua_subscription_alarms_conditions.c:2675]: can't this be done on the stack. the life-time of the variant is well-defined and short. no need to use the heap?

so also the other cases

**Comment by opcua-tsn-team-kalycito** [comment]
@jpfr, Addressed the review comment and rebased to master. Can you have a look and merge the PR?

**Comment by opcua-tsn-team-kalycito** [comment]
Hi @jpfr ,
Review comments are addressed and rebased to latest master, Can you look into it and approve the same.

**Comment by jpfr** [comment]
Thanks for the update.
The code is looking good.
But we should keep the separation between core library and os-dependent plugins intact.

To get the expiration date of a certificate, please add this functionality to the `UA_CertificateVerification` structure in
`/include/open62541/pki.h`.

Your mbedtls implementation can be copied verbatim to the corresponding plugin.
The pointer can be left as NULL for the OpenSSL implementation.
With an error message that the date cannot be checked.

Then your A&C code can use the certificate verification plugin that is globally registered in the server config.

**Comment by opcua-tsn-team-kalycito** [comment]
Hi @jpfr ,
Review comments are addressed, Can you look into it and merge the same?

File: include/open62541/server.h [hunk 2]
```diff
@@ -1738,6 +1738,31 @@ UA_Server_deleteCondition(UA_Server *server,
...
+*
+* @param server The server object
+* @param conditionId The NodeId of the node representation of the Condition Instance
+* @param limitValue The value from the trigger node
+*/
+UA_StatusCode UA_EXPORT
+UA_Server_setLimitState(UA_Server *server, const UA_NodeId conditionId,
+                        UA_Double limitValue);
+
+#if defined(UA_ENABLE_ENCRYPTION) && \
```
Comment (jpfr) [inline include/open62541/server.h:1752]: That's too much ifedeffery.

UA_Server_setExpirationDate can be always defined.
If encryption is disabled (or the pointer inside certificateVerification is NULL), log and return an error code.

File: src/server/ua_subscription_alarms_conditions.c [hunk 3]
```diff
@@ -2638,4 +2718,175 @@ UA_StatusCode UA_Server_deleteCondition(UA_Server *server, const UA_NodeId condi
     /* Delete from address space */
     return UA_Server_deleteNode(server, condition, true);
 }
-#endif /* UA_ENABLE_SUBSCRIPTIONS_ALARMS_CONDITIONS */
+
+static UA_StatusCode getLowLimit(UA_Server *server, UA_NodeId conditionId, UA_Double *lowLimit) {
```
Comment (jpfr) [inline src/server/ua_subscription_alarms_conditions.c:2722]: Shorten the lines.
Add a break after `static UA_StatusCode` as is done in the rest of the code.
The same below.

File: src/server/ua_subscription_alarms_conditions.c [hunk 4]
```diff
@@ -2638,4 +2718,175 @@ UA_StatusCode UA_Server_deleteCondition(UA_Server *server, const UA_NodeId condi
...
+    }
+
+    /* Write the expiry date to the propery of the condition instance */
+    retval = UA_Server_writeObjectProperty_scalar(server, conditionId,
+                                                  fieldExpirationDateQN,
+                                                  &getExpiryDateAndTime,
+                                                  &UA_TYPES[UA_TYPES_DATETIME]);
+    return retval;
+}
+#endif//UA_ENABLE_ENCRYPTION
```
Comment (jpfr) [inline src/server/ua_subscription_alarms_conditions.c:2891]: Use C-comments /* */ instead of C++ comments //

**Review by jpfr — COMMENTED** [review]
Nice!
This is nearly ready.
Just a few simplifications with less #ifdef and this can be merged.

**Comment by opcua-tsn-team-kalycito** [comment]
Hi @jpfr ,
Latest review comments are addressed, Can you look into and merge the same?

**Comment by jpfr** [comment]
Nice! Merged.

**Pull request merged**